feat: redesign modal checkout

miguelpeixe opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-theme/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

1205234045751551-as-1205933702651188

Implements a redesigned and new 2-step strategy for the modal checkout.

The modal checkout class no longer performs form validation, it now uses `woocommerce_checkout_update_totals` so  [it validates through Woo without having to place the order](https://github.com/woocommerce/woocommerce/blob/798e79f7b1bb5cd3db376de8712ad122d0094112/plugins/woocommerce/includes/class-wc-checkout.php#L1262). By validating through Woo, the modal checkout should natively support any custom field combination Woo and its extensions may use. This new strategy also uses XHR instead of a regular form submission, which should provide a better and faster experience for the reader.

To also increase support for core Woo checkout functionality, the functional aspects of the modal checkout were moved from the template file to [hooks](https://github.com/Automattic/newspack-blocks/blob/7fd2af31749c8570148b7a0c5294a2443e0e8014/includes/class-modal-checkout.php#L52-L54). This allows the template file to be modified/updated without impacting the requirements that allow the checkout to function in a modal. The new `src/modal-checkout/templates/form-checkout.php` is now intentionally very similar to Woo's original template file.

A significant part of the work here has been around `src/modal-checkout/index.js`, to allow the new 2-step approach to work as the design proposes. Some important parts are:

 - When in the first step (editing billing/shipping fields), [the original form event handlers are replaced](https://github.com/Automattic/newspack-blocks/blob/635a685a22878fde6649cc4cb2aef8b96e93ef5d/src/modal-checkout/index.js#L180-L185) so the form submission performs [our ajax request](https://github.com/Automattic/newspack-blocks/blob/635a685a22878fde6649cc4cb2aef8b96e93ef5d/src/modal-checkout/index.js#L286) (which adds the `woocommerce_checkout_update_totals` parameter).
 - [A custom error handler](https://github.com/Automattic/newspack-blocks/blob/7fd2af31749c8570148b7a0c5294a2443e0e8014/src/modal-checkout/index.js#L84-L89) for the 1st step identifies whether the incoming errors are field-related or generic and appends the error message to the field or at the top accordingly.

### Known issues

- The coupon form is inconsistent. When applying a coupon the form disappears, which can be considered fine because it reappears after removing the coupon. It's not ok when applying an inexistent coupon, which shows you an error message and the coupon form disappears, becoming irrecoverable. (PR: https://github.com/Automattic/newspack-blocks/pull/1629)
- Linter is accusing a CSS kebab-case issue even though the line has `/* stylelint-disable-line */`. Not sure how to deal with this.

### Subscription

<img width="647" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/87c99af8-4153-455b-a9e5-98e4a8c5235b">

<img width="638" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/9e1d0d58-2488-4777-8032-6b55122097a7">

### One-time donation

<img width="661" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/e8017d85-1156-4378-a85b-47161b714777">

### Recurring donation

<img width="651" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/b2da7fe1-0e14-419b-bff1-a4a877ac7604">

### Recurring donation (known customer with known card)

<img width="651" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/629a198a-ea16-4bb4-83c0-cc5882f1ea60">

### Shippable product

<img width="649" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/f7a2e5b6-ed99-4492-806c-0b9cb193c9b8">

### Field validation

<img width="641" alt="image" src="https://github.com/Automattic/newspack-blocks/assets/820752/202d7b70-f5f7-4a96-9b33-a1e8b84ab25a">

### How to test the changes in this Pull Request:

1. Checkout this branch and https://github.com/Automattic/newspack-plugin/pull/2805
2. If you don't have any, setup a shipping zone in Woo:
    1. Visit WooCommerce -> Settings -> Shipping
    2. Click "Add shipping zone"
    3. Give it a name (the country name)
    4. Set the "Zone region" to your country
    5. Click "Add shipping method" and leave it as "Flat rate"
    6. Save changes
3. Make sure you have coupons enabled and have a valid coupon
7. Create a product that can be shipped
8. Create a regular custom subscription
9. Create a page with a Checkout Button block for each of the created products
10. Also add a Donate Block to the page
11. Visit the page, click to buy the shippable product
12. Confirm billing and shipping fields render as expected
13. Attempt to "Continue" without entering all required fields
14. Confirm validation errors render for each field
15. Enter a zipcode that you know to be an invalid format
16. Confirm it errors with "Billing ZIP Code is not a valid postcode / ZIP."
17. Now enter valid fields and confirm you are able to go to the next step
18. Confirm the second step renders a summary of the "Billing details" and "Shipping details"
19. Click "Back" and confirm you are able to edit the fields again
20. Scroll down and toggle "Ship to a different address?"
21. Modify the address and click "Continue"
22. Confirm the summary reflects the different address
23. Apply the coupon and confirm the order review table renders with the coupon information
24. Click to remove the coupon and confirm the order review table is no longer visible
25. Re-apply the coupon so we can confirm the coupon persists while placing the order
26. Submit and confirm the order is placed with the applied coupon
27. Leave the modal and click to purchase the custom subscription
28. Confirm the billing fields reflect the selection from "Newspack -> Reader Revenue"
29. Go through the entire flow and confirm the order is placed without issues
30. Leave the modal and click to donate
31. In the second step, confirm the coupon form is not available
32. Confirm the button to place the order says "Donate now" instead of "Complete transaction"
33. Click to donate and confirm the order is placed

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

miguelpeixe committed:
feat: redesign modal checkout

miguelpeixe committed:
fix: saved methods spacing

miguelpeixe committed:
fix: style updates and start custom coupon form

miguelpeixe committed:
feat: start front-end 2-step

miguelpeixe committed:
fix: style tweaks and button label

miguelpeixe committed:
feat: animate modal in

dkoo commented:
@miguelpeixe let's base redesign work from the [`epic/ras-acc` branch](https://github.com/Automattic/newspack-blocks/tree/epic/ras-acc) so we can control the release.

miguelpeixe committed:
fix: style tweak

miguelpeixe committed:
fix: hide radio if only one payment method

miguelpeixe committed:
fix: tweak spacing for payment form

miguelpeixe committed:
fix: variation modal and mobile tweaks

miguelpeixe committed:
fix: temporarily remove 2-step and ux tweaks

miguelpeixe committed:
feat: custom shipping address option

miguelpeixe committed:
chore: lint

miguelpeixe committed:
chore: clode cleanup

miguelpeixe committed:
chore: restore commented-out code

miguelpeixe committed:
fix: tweak address field margin

miguelpeixe committed:
chore: readability

miguelpeixe committed:
fix: remove donation text from order button

miguelpeixe committed:
fix: do shortcode instead of loading page content

miguelpeixe committed:
Merge branch 'epic/ras-acc' into feat/modal-checkout-redesign

miguelpeixe committed:
fix: select2 styles

miguelpeixe committed:
fix: more select2 style tweaks

miguelpeixe committed:
Merge branch 'epic/ras-acc' into feat/modal-checkout-redesign

miguelpeixe committed:
feat: preliminary work on 2-step with validation

miguelpeixe committed:
fix: restore original handlers

miguelpeixe committed:
fix: clear messages and scroll top

miguelpeixe committed:
fix: better handling of input errors

miguelpeixe committed:
fix: remove red input shadow

miguelpeixe committed:
fix: handle back button and improve js logic

miguelpeixe committed:
chore: code cleanup

miguelpeixe committed:
fix: remove iframe handling of inner forms

miguelpeixe committed:
feat: implement initial validation

miguelpeixe committed:
feat: build checkout details

miguelpeixe committed:
fix: clear details

miguelpeixe committed:
feat: ux and performance tweaks

miguelpeixe committed:
fix: resolve spinner if container is not available

miguelpeixe committed:
fix: remove block notices

miguelpeixe committed:
chore: add todo

miguelpeixe committed:
chore: remove unused code

miguelpeixe committed:
chore: remove unused code

miguelpeixe committed:
chore: remove unused code

miguelpeixe committed:
chore: use helper method

miguelpeixe committed:
fix: handle new error format

miguelpeixe committed:
Merge branch 'epic/ras-acc' into feat/modal-checkout-redesign

miguelpeixe committed:
fix: better handle notices and shipping details

miguelpeixe committed:
chore: update line docs

miguelpeixe committed:
chore: update inline docs

miguelpeixe committed:
chore: code cleanup and inline docs

miguelpeixe committed:
chore: remove superfluous phpcs ignore

miguelpeixe committed:
chore: improve readability

miguelpeixe committed:
feat: focus first field on editing

miguelpeixe committed:
chore: improve readability

miguelpeixe committed:
fix: check coupon form

dkoo commented:
Getting started with testing and reviewing this (finally! sorry for the delay). But I have an answer to one of your questions!

 > Linter is accusing a CSS kebab-case issue even though the line has /* stylelint-disable-line */. Not sure how to deal with this.

I ran into this same issue on #1626. The fix is to put both CSS selectors on the same line, like [this](https://github.com/Automattic/newspack-blocks/pull/1626/files#diff-93359786372d1c41af813037af697f33c6a6da8d4a48622a3c18d8e6f26f1cd7).

```css
#checkout_continue, #place_order { /* stylelint-disable-line */
```

dkoo reviewed (changes requested):
I went through all testing steps and everything is mostly working! 🙌 

Before I do a more thorough code review I ran into a couple of minor things so far:

- When checking the "I’d like to cover the 3.6% transaction fee" option, the transaction total is updated at the top of the modal, but not in the Transaction Details summary's total (the order amount is correct once placed, though).
- If taxes are not enabled and no coupons are applied, when purchasing a shippable product with either multiple available shipping methods and/or only paid shipping methods, the Transaction Details summary is not shown. This results in a strange modal state where the totals don't match up (this _should_ be fixed by #1626, though, so maybe there's nothing needed for this PR):

<img width="643" alt="Screenshot 2023-12-18 at 5 28 32 PM" src="https://github.com/Automattic/newspack-blocks/assets/2230142/90403914-8035-4dde-9b0a-c7deefc53be3">


More testing and a line-by-line code review to come!

miguelpeixe committed:
fix: show order details if requires shipping

miguelpeixe committed:
chore: fix linter

miguelpeixe commented:
Thank you for taking this review, @dkoo!

> When checking the "I’d like to cover the 3.6% transaction fee" option, the transaction total is updated at the top of the modal, but not in the Transaction Details summary's total (the order amount is correct once placed, though).

Fixed in the https://github.com/Automattic/newspack-plugin/pull/2805 (https://github.com/Automattic/newspack-plugin/pull/2805/commits/6ff7dd04fba6ab130fba0bdf443bd03bf31e4219).

> If taxes are not enabled and no coupons are applied, when purchasing a shippable product with either multiple available shipping methods and/or only paid shipping methods, the Transaction Details summary is not shown.

I didn't find the mentioned PR tweak the rules for when to render the order review table, so I added in e00793fe9f5b991c7b8014e2b2724e72f0d69a2b

> I ran into this same issue on https://github.com/Automattic/newspack-blocks/pull/1626. The fix is to put both CSS selectors on the same line, like [this](https://github.com/Automattic/newspack-blocks/pull/1626/files#diff-93359786372d1c41af813037af697f33c6a6da8d4a48622a3c18d8e6f26f1cd7).

😌 updated in fc748a28f2c435d38546d6e2c161ef112b20bf48

miguelpeixe commented:
Update regarding the cover fee issue: https://github.com/Automattic/newspack-plugin/pull/2805#issuecomment-1862850079

dkoo commented:
> The mentioned PR doesn't seem to tweak the rules for when to render the order review table, so I added in https://github.com/Automattic/newspack-blocks/commit/e00793fe9f5b991c7b8014e2b2724e72f0d69a2b

It does: https://github.com/Automattic/newspack-blocks/pull/1626/files#diff-74a92f16c32d2a94751337aaff7d24de10881675f81809bae9d09e79cc47e10eR645

But it uses `$cart->needs_shipping_address()` instead of just `$cart->needs_shipping()`, so maybe it would need to be tweaked.

miguelpeixe commented:
Ah, I missed that. I think my brain was looking for a one-liner 😅 . I'll revert my change and allow your PR to handle those requirements.

miguelpeixe committed:
Revert "fix: show order details if requires shipping"

This reverts commit e00793fe9f5b991c7b8014e2b2724e72f0d69a2b.

dkoo reviewed (approved):
Looking good! Issues and solutions in separate PRs noted. I think this is a fantastic foundation for the redesigned modal checkout experience, and it lays the foundation for a lot of great work to come!

Tested with one-time donations, recurring donations, non-donation purchases (one-time and subscriptions), variable subscriptions, and shippable products. Couldn't break it.

miguelpeixe commented on includes/class-modal-checkout.php:
```diff
@@ -967,10 +883,113 @@ public static function is_modal_checkout() {
+						echo apply_filters( 'woocommerce_cart_item_name', $_product->get_name(), $cart_item, $cart_item_key ) . ': '; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+						echo wc_get_formatted_cart_item_data( $cart_item ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+						?>
+						<?php echo apply_filters( 'woocommerce_cart_item_subtotal', $cart->get_product_subtotal( $_product, $cart_item['quantity'] ), $cart_item, $cart_item_key ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
```
> Maybe a simple way to deal with this is to add a class for the donation checkout container, and add display: none to .subscription-details class name?

Good idea! I believe we should tackle this separately in the plugin's Donation class. I want to avoid adding donation-specific logic here as much as we can for better maintainability.

There's more to this line that I'd like to discuss. It's currently not providing the total order value, only the product item price. I think we should change that so it can account for additional values (coupon, fees (https://github.com/Automattic/newspack-plugin/pull/2820) and taxes)

dkoo replied:
Very cool—love the idea of using cart fees to handle the transaction fees separately!
